python/pylibcudf/pylibcudf/strings/replace_re.pyx

```diff
@@ -0,0 +1,131 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.
+from cython.operator cimport dereference
+from libcpp.memory cimport unique_ptr
+from libcpp.string cimport string
+from libcpp.utility cimport move
+from libcpp.vector cimport vector
+from pylibcudf.column cimport Column
+from pylibcudf.libcudf.column.column cimport column
+from pylibcudf.libcudf.scalar.scalar cimport string_scalar
+from pylibcudf.libcudf.strings cimport replace_re as cpp_replace_re
+from pylibcudf.libcudf.types cimport size_type
+from pylibcudf.scalar cimport Scalar
+from pylibcudf.strings.regex_flags cimport regex_flags
+from pylibcudf.strings.regex_program cimport RegexProgram
+
+
+cpdef Column replace_re(
+    Column input,
+    Patterns patterns,
+    Replacement replacement,
+    size_type max_replace_count=-1,
+    regex_flags flags=regex_flags.DEFAULT,
+):
+    """
+    For each string, replaces any character sequence matching the given patterns
+    with the provided replacement.
+
+    For details, see :cpp:func:`cudf::strings::replace_re`
+
+    Parameters
+    ----------
+    input : Column
+        Strings instance for this operation.
+
+    patterns: RegexProgram or list[str]
+        If RegexProgram, the regex to match to each string.
+        If list[str], a list of regex strings to search within each string.
+
+    replacement : Scalar or Column
+        If Scalar, the string used to replace the matched sequence in each string.
+        ``patterns`` must be a RegexProgram.
+        If Column, the strings used for replacement.
+        ``patterns`` must be a list[str].
+
+    max_replace_count : int
+        The maximum number of times to replace the matched pattern
+        within each string. ``patterns`` must be a RegexProgram.
+        Default replaces every substring that is matched.
+
+    flags : RegexFlags
+        Regex flags for interpreting special characters in the patterns.
+        ``patterns`` must be a list[str]
+
+    Returns
+    -------
+    Column
+        New strings column
+    """
+    cdef unique_ptr[column] c_result
+    cdef vector[string] c_patterns
+
+    if patterns is RegexProgram and replacement is Scalar:
+        with nogil:
+            c_result = move(
+                cpp_replace_re.replace_re(
+                    input.view(),
+                    patterns.c_obj.get()[0],
+                    dereference(<string_scalar*>(replacement.get())),
```

Comment:
I'm getting a compilation error here

```bash
Casting temporary Python object to non-numeric non-Python type
```

Would appreciate a second set of eyes to see if I missed anything obvious

Reply:
Try adding an explicit cast 
```suggestion
                    dereference(<string_scalar*>((<Scalar> replacement).get())),
```
If that works, I'd try and work up an MWE and submit a bug report to Cython if needed. This definitely seems like it should work, but I think the underlying cause of the error is that it doesn't know the type of `replacement` well enough to recognize `replacement.get()` as a typed call.

Reply:
Thanks that cast worked (I had to generally, unfortunately cast after the type narrowing of the fused types). 

When I casted `(cdef regex_program) c_regex_pattern = (<RegexProgram>patterns).c_obj.get()[0]`, I do get this compilation error though

```bash
      /home/nfs/mroeschke/mroeschke-cudf/python/pylibcudf/build/cp312-cp312-linux_x86_64/pylibcudf/strings/replace_re.cxx: In function '__pyx_obj_9pylibcudf_6column_Column* __pyx_fuse_0_0__pyx_f_9pylibcudf_7strings_10replace_re_replace_re(__pyx_obj_9pylibcudf_6column_Column*, __pyx_obj_9pylibcudf_7strings_13regex_program_RegexProgram*, __pyx_obj_9pylibcudf_6column_Column*, int, __pyx_fuse_0_0__pyx_opt_args_9pylibcudf_7strings_10replace_re_replace_re*)':
      /home/nfs/mroeschke/mroeschke-cudf/python/pylibcudf/build/cp312-cp312-linux_x86_64/pylibcudf/strings/replace_re.cxx:21890:32: error: use of deleted function 'cudf::strings::regex_program::regex_program()'
      21890 |   cudf::strings::regex_program __pyx_v_c_regex_pattern;
            |                                ^~~~~~~~~~~~~~~~~~~~~~~
      In file included from /home/nfs/mroeschke/mroeschke-cudf/python/pylibcudf/build/cp312-cp312-linux_x86_64/pylibcudf/strings/replace_re.cxx:1277:
      /home/nfs/mroeschke/miniforge3/envs/cudf-dev/include/cudf/strings/regex/regex_program.hpp:57:3: note: declared here
         57 |   regex_program() = delete;
            |   ^~~~~~~~~~~~~
      /home/nfs/mroeschke/mroeschke-cudf/python/pylibcudf/build/cp312-cp312-linux_x86_64/pylibcudf/strings/replace_re.cxx:21955:64: error: use of deleted function 'cudf::strings::regex_program& cudf::strings::regex_program::operator=(const cudf::strings::regex_program&)'
      21955 |     __pyx_v_c_regex_pattern = (__pyx_v_patterns->c_obj.get()[0]);
            |                                                                ^
      In file included from /home/nfs/mroeschke/mroeschke-cudf/python/pylibcudf/build/cp312-cp312-linux_x86_64/pylibcudf/strings/replace_re.cxx:1277:
      /home/nfs/mroeschke/miniforge3/envs/cudf-dev/include/cudf/strings/regex/regex_program.hpp:40:8: note: 'cudf::strings::regex_program& cudf::strings::regex_program::operator=(const cudf::strings::regex_program&)' is implicitly declared as deleted because 'cudf::strings::regex_program' declares a move constructor or move assignment operator
         40 | struct regex_program {
            |        ^~~~~~~~~~~~~
      /home/nfs/mroeschke/mroeschke-cudf/python/pylibcudf/build/cp312-cp312-linux_x86_64/pylibcudf/strings/replace_re.cxx: In function '__pyx_obj_9pylibcudf_6column_Column* __pyx_fuse_0_1__pyx_f_9pylibcudf_7strings_10replace_re_replace_re(__pyx_obj_9pylibcudf_6column_Column*, __pyx_obj_9pylibcudf_7strings_13regex_program_RegexProgram*, __pyx_obj_9pylibcudf_6scalar_Scalar*, int, __pyx_fuse_0_1__pyx_opt_args_9pylibcudf_7strings_10replace_re_replace_re*)':
      /home/nfs/mroeschke/mroeschke-cudf/python/pylibcudf/build/cp312-cp312-linux_x86_64/pylibcudf/strings/replace_re.cxx:22538:32: error: use of deleted function 'cudf::strings::regex_program::regex_program()'
      22538 |   cudf::strings::regex_program __pyx_v_c_regex_pattern;
            |                                ^~~~~~~~~~~~~~~~~~~~~~~
      In file included from /home/nfs/mroeschke/mroeschke-cudf/python/pylibcudf/build/cp312-cp312-linux_x86_64/pylibcudf/strings/replace_re.cxx:1277:
      /home/nfs/mroeschke/miniforge3/envs/cudf-dev/include/cudf/strings/regex/regex_program.hpp:57:3: note: declared here
         57 |   regex_program() = delete;
            |   ^~~~~~~~~~~~~
      /home/nfs/mroeschke/mroeschke-cudf/python/pylibcudf/build/cp312-cp312-linux_x86_64/pylibcudf/strings/replace_re.cxx:22603:64: error: use of deleted function 'cudf::strings::regex_program& cudf::strings::regex_program::operator=(const cudf::strings::regex_program&)'
      22603 |     __pyx_v_c_regex_pattern = (__pyx_v_patterns->c_obj.get()[0]);
            |                                                                ^
      /home/nfs/mroeschke/mroeschke-cudf/python/pylibcudf/build/cp312-cp312-linux_x86_64/pylibcudf/strings/replace_re.cxx:22759:35: error: no match for 'operator=' (operand types are 'cudf::column_view' and 'PyObject*' {aka '_object*'})
      22759 |       __pyx_v_c_replacement_col = __pyx_t_5;
```

Reply:
You're running into a common problem in Cython's C++ integration. You can see it in [the docs](https://cython.readthedocs.io/en/latest/src/userguide/wrapping_CPlusPlus.html#declare-a-var-with-the-wrapped-c-class), specifically this:

> It’s also possible to declare a stack allocated object, as long as it has a “default” constructor:

The problem is that if you have Cython that looks like this
```
cdef MyType t = make_t()
```

The C++ Cython generates will always look like this:
```
MyType t;
// Do other things
t = make_t(...)
```

The line `MyType t;` is only valid if `MyType` is default-constructible, which is explicitly not the case for `regex_program` because the `regex_program()` default constructor [is deleted](https://github.com/rapidsai/cudf/blob/branch-24.12/cpp/include/cudf/strings/regex/regex_program.hpp#L57).

The typical workaround for this is to use a pointer to MyType instead of a stack-allocated variable. It's not as nice since you have to manage the lifetime (ideally by using a smart pointer) and you end up with heap overhead, but that's the price you pay.

Reply:
If possible I would definitely try and write up a bug report for Cython devs of the above. I do think that code _should_ work.

Reply:
I think the problem was we were comparing the argument names and not the types like `if patterns is RegexProgram` and not `if Patterns is RegexProgram`. We should be doing the latter.

Reply:
Thanks so much for catching this!

Reply:
Good eye! I missed that when reviewing.